package.json

```diff
@@ -18,7 +22,16 @@
     "webpack-cli": "^4.2.0"
   },
   "devDependencies": {
+    "elm": "^0.19.1-5",
+    "elm-format": "^0.8.5",
     "elm-minify": "^2.0.4",
+    "elm-test": "^0.19.1-revision6",
+    "husky": "^5.1.3",
+    "lint-staged": "^10.5.4",
+    "serve": "^11.3.2",
     "webpack-dev-server": "3.11.0"
+  },
+  "lint-staged": {
+    "*.elm": "npm run format --"
```

Comment:
Is there a way to prevent a commit from occurring if the formatting is incorrect (as opposed to autoformatting on commit)?

I personally do not like it when tools change my code on save or on commit, but I would appreciate if husky / lint staged prevented me from making a commit that failed format / style rules.

Reply:
I disagree with that. I know it can be a bit disorienting but the point of elm-format is to not have to worry about formatting. This commit hook abstracts that out in a way that the dev doesn't have to worry about the whether they have an extra line somewhere. Just stopping the dev on commit will be inefficient as the dev will have to run elm-format manually, anyway.
Ideally, all devs should also have elm auto-format configured in their editors that way this hook won't have make any changes.

Reply:
i think checking for formatting is more of a ci tools responsibility, tools like husky and others that modify git hooks to block commits can be easily bypassed using `--no-verify` flag. They are also time wasting, since they run checks on each commit

Since it's open source, it is easy and cost less to setup a github action that checks for code formatting on each pull request. 
We can migrate to github actions from circle ci (as far as i know circle ci is no longer free for open source or they some limits)

Reply:
@aszenz. This pr is not trying to do the CI's job. The husky hook's purpose is to allow developers to write and commit code without worrying about its formatting. The cycle of having the CI stop you if you formatted your code incorrectly, then running the formatter manually on your branch, then making a new commit for formatting and then finally pushing it back to get the CI happy is an inefficiency.

I do agree that people can bypass this so we should have a CI but I think having the husky hook for auto-formatting is a much better dev-ex than having the CI stop you every time.

Reply:
This convo is full of opinions. If a developer wants a particular set up on their own machine, then they're free to do so. But to impose a particular way of doing things on a developer is not my style unless there's good reason to do so. And I just don't see a compelling enough argument here.

Would it be possible to make this opt-in with a simple bash script within the project?

Reply:
I don't agree with the choice since having that hook is effortless, never gets in the way so there is no risk associated with it. I'll remove it but let it be known to everyone in the kingdom that I do so with a pout on my face.

Reply:
glad we could agree to disagree (;

i personally also don't worry about formatting by adding a `.git/hooks/pre-commit.sh` file which does the formatting, 
also have auto format on save enabled in my editor.

here's a script you could use for pre-commit

```bash
#!/usr/bin/env bash
FORMAT="npm format"
git status --porcelain | grep -e '^[AM]\(.*\).elm$' | cut -c 3- | while read line; do
      $FORMAT "$line";
      git add "$line";
done
```

Reply:
@aszenz, can we merge this? I'll create a new PR applying formatting across the project.